### PR TITLE
Properly handle stdin forwarding

### PR DESCRIPTION
**Pull request by rhc54**
Signed-off-by: Ralph Castain <rhc@pmix.org>

### CODE REVIEW

**Comment by rhc54** [comment]
bot:ibm:prrte:retest

**Comment by jjhursey** [comment]
bot:ibm:prrte:retest

**Comment by rhc54** [comment]
I'm afraid this is a real error - I'm looking into it today

**Comment by rhc54** [comment]
@jjhursey I can't seem to reproduce this error on my VM setup. I only have 3 nodes currently replicated, but even running 10ppn passes with no problem. Can you perhaps take a closer peek at what is happening on the CI? I'm happy to fix it, though I'm puzzled as to why changes to the stdin handling would affect stdout - there shouldn't be a connection unless we are shutting down too early.

**Comment by jjhursey** [comment]
I can try to take a look later. One thing I've noticed in CI is that it essentially does `prun < /dev/null` (it comes from the Jenkins higher level it seems) so maybe that's related - or something to try to reproduce locally.

**Comment by rhc54** [comment]
At least on the Mac (not in the VM), it works fine with `prun < /dev/null`. I'll check in the VM next

**Comment by rhc54** [comment]
In the VM, `prun` immediately segfaults if I tie it to `/dev/null`. Unfortunately, the VM is no longer providing me with corefiles - no idea when/why that changed. Are you able to get them? If so, maybe we should resync through your repo as I must have messed something up locally. All I get is a "backtrace file":
```text
$ cat  prun.80s-40,rhc-node01.btr
prun:40 terminated with signal 11 at PC=0 SP=7fda58702c08.  Backtrace:
/lib64/libinfinipath.so.4(+0x419c)[0x7fda559c419c]
/lib64/libpthread.so.0(+0xf630)[0x7fda5c3ba630]
$
```

**Comment by jjhursey** [comment]
I hit the following (probably unrelated) segv this morning while playing with this (bad binary name):
```
shell$ prte --no-ready-msg --hostfile /workspace/hostfile.txt < /dev/null
shell$ prun --map-by ppr:5:node ./hostfile.txt < /dev/null | wc -l
prte: base/state_base_fns.c:85: prte_state_base_activate_job_state: Assertion `PRTE_OBJ_MAGIC_ID == ((prte_object_t *) (jdata))->obj_magic_id' failed.
--------------------------------------------------------------------------
prte was unable to launch the specified application as it could not access
or execute an executable:

Executable: ./hostfile.txt
Node: 9dbd12fd420b

while attempting to start process rank 0.
--------------------------------------------------------------------------

0
[1]+  Aborted                 (core dumped) prte --no-ready-msg --hostfile /workspace/hostfile.txt < /dev/null
```

```
[jjhursey@9dbd12fd420b (docker) hello_world] gdb prte --core core.62422 
(gdb) bt
#0  0x00007fffa3c84308 in raise () from /usr/lib64/libc.so.6
#1  0x00007fffa3c64484 in abort () from /usr/lib64/libc.so.6
#2  0x00007fffa3c778d0 in __assert_fail_base () from /usr/lib64/libc.so.6
#3  0x00007fffa3c77974 in __assert_fail () from /usr/lib64/libc.so.6
#4  0x00007fffa476a2ac in prte_state_base_activate_job_state (jdata=0x10030d65d00, state=35) at base/state_base_fns.c:85
#5  0x00007fffa3aea674 in dvm_notify (sd=-1, args=4, cbdata=0x7fff4c002eb0) at state_dvm.c:921
#6  0x00007fffa42e9e24 in event_process_active_single_queue (base=base@entry=0x10030a92c90, activeq=0x10030a88910, 
    max_to_process=max_to_process@entry=2147483647, endtime=endtime@entry=0x0) at event.c:1691
#7  0x00007fffa42ea9b0 in event_process_active (base=0x10030a92c90) at event.c:1783
#8  event_base_loop (base=0x10030a92c90, flags=<optimized out>) at event.c:2006
#9  0x000000001000aa30 in prte (argc=4, argv=0x7fffe3c05c98) at prte.c:1122
#10 0x00000000100039d4 in main (argc=4, argv=0x7fffe3c05c98) at main.c:13
(gdb) up
#1  0x00007fffa3c64484 in abort () from /usr/lib64/libc.so.6
(gdb) 
#2  0x00007fffa3c778d0 in __assert_fail_base () from /usr/lib64/libc.so.6
(gdb) 
#3  0x00007fffa3c77974 in __assert_fail () from /usr/lib64/libc.so.6
(gdb) 
#4  0x00007fffa476a2ac in prte_state_base_activate_job_state (jdata=0x10030d65d00, state=35) at base/state_base_fns.c:85
85	                PRTE_RETAIN(jdata);
(gdb) l
80	            }
81	            caddy = PRTE_NEW(prte_state_caddy_t);
82	            if (NULL != jdata) {
83	                caddy->jdata = jdata;
84	                caddy->job_state = state;
85	                PRTE_RETAIN(jdata);
86	            }
87	            PRTE_THREADSHIFT(caddy, prte_event_base, s->cbfunc, s->priority);
88	            return;
89	        }
(gdb) p jdata
$1 = (prte_job_t *) 0x10030d65d00
(gdb) p *jdata
$2 = {
  super = {
    super = {
      obj_magic_id = 1100330980640, 
      obj_class = 0x10030d614f0, 
      obj_reference_count = 0, 
      cls_init_file_name = 0x0, 
      cls_init_lineno = 111
    }, 
    prte_list_next = 0x0, 
    prte_list_prev = 0x0, 
    item_free = 1, 
    prte_list_item_refcount = 0, 
    prte_list_item_belong_to = 0x0
  }, 
  exit_code = 76, 
  personality = 0x10030c15c80, 
  nspace = "prte-9dbd12fd420b-62422@22", '\000' <repeats 229 times>, 
  index = 2, 
  offset = 500, 
  apps = 0x0, 
  num_apps = 1, 
  stdin_target = 0, 
  total_slots_alloc = 200, 
  num_procs = 50, 
  procs = 0x0, 
  map = 0x0, 
  bookmark = 0x10030cba4a0, 
  bkmark_obj = 4294967295, 
  state = 35, 
  num_mapped = 0, 
  num_launched = 0, 
  num_reported = 0, 
  num_terminated = 0, 
  num_daemons_reported = 1, 
  num_ready_for_debug = 0, 
  originator = {
    nspace = "prte-9dbd12fd420b-62422@0", '\000' <repeats 230 times>, 
    rank = 0
  }, 
  num_local_procs = 5, 
  flags = 16424, 
  attributes = {
    super = {
      obj_magic_id = 0, 
--Type <RET> for more, q to quit, c to continue without paging--
      obj_class = 0x7fffa47d14f8 <prte_list_t_class>, 
      obj_reference_count = 1, 
      cls_init_file_name = 0x7fffa4773450 "runtime/prte_globals.c", 
      cls_init_lineno = 572
    }, 
    prte_list_sentinel = {
      super = {
        obj_magic_id = 16046253926196952813, 
        obj_class = 0x7fffa47d14b8 <prte_list_item_t_class>, 
        obj_reference_count = 1, 
        cls_init_file_name = 0x7fffa4771608 "class/prte_list.c", 
        cls_init_lineno = 90
      }, 
      prte_list_next = 0x10030d65fe8, 
      prte_list_prev = 0x10030d65fe8, 
      item_free = 1, 
      prte_list_item_refcount = 1, 
      prte_list_item_belong_to = 0x10030d65fc0
    }, 
    prte_list_length = 0
  }, 
  launch_msg = {
    base_ptr = 0x0, 
    pack_ptr = 0x0, 
    unpack_ptr = 0x0, 
    bytes_allocated = 0, 
    bytes_used = 0
  }, 
  children = {
    super = {
      obj_magic_id = 0, 
      obj_class = 0x7fffa47d14f8 <prte_list_t_class>, 
      obj_reference_count = 1, 
      cls_init_file_name = 0x7fffa4773450 "runtime/prte_globals.c", 
      cls_init_lineno = 581
    }, 
    prte_list_sentinel = {
      super = {
        obj_magic_id = 16046253926196952813, 
        obj_class = 0x7fffa47d14b8 <prte_list_item_t_class>, 
        obj_reference_count = 1, 
        cls_init_file_name = 0x7fffa4771608 "class/prte_list.c", 
        cls_init_lineno = 90
      }, 
      prte_list_next = 0x10030d66088, 
--Type <RET> for more, q to quit, c to continue without paging--
      prte_list_prev = 0x10030d66088, 
      item_free = 1, 
      prte_list_item_refcount = 1, 
      prte_list_item_belong_to = 0x10030d66060
    }, 
    prte_list_length = 0
  }, 
  launcher = '\000' <repeats 248 times>...
}
```

**Comment by jjhursey** [comment]
With the latest HEAD of PMIx and PRRTE everything looks good. I'm trying now with this branch.

**Comment by rhc54** [comment]
Yeah, HEAD of master on PMIx doesn't actually activate the pull on stdin - I forgot I hadn't implemented it yet. 🤷‍♂️ 

I'll take a look at that other segv in a minute.

**Comment by jjhursey** [comment]
Reproduced segv with core using:
```
prte --no-ready-msg --hostfile /workspace/hostfile.txt &
prun --map-by ppr:5:node hostname < /dev/null 
```

```
[jjhursey@9dbd12fd420b (docker) hello_world] gdb prun --core core.140788 
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x00007fffb4e57318 in pmix_iof_read_local_handler (unusedfd=-1, event=1, cbdata=0x10031f89840) at common/pmix_iof.c:1232
#2  0x00007fffb4c59e24 in event_process_active_single_queue (base=base@entry=0x10031ef0a00, activeq=0x10031ef0ed0, 
    max_to_process=max_to_process@entry=2147483647, endtime=endtime@entry=0x0) at event.c:1691
#3  0x00007fffb4c5a9b0 in event_process_active (base=0x10031ef0a00) at event.c:1783
#4  event_base_loop (base=0x10031ef0a00, flags=<optimized out>) at event.c:2006
#5  0x00007fffb4e1cdc4 in progress_engine (obj=0x10031ef0990) at runtime/pmix_progress_threads.c:235
#6  0x00007fffb47d8ba0 in start_thread () from /usr/lib64/libpthread.so.0
#7  0x00007fffb46e1258 in clone () from /usr/lib64/libc.so.6
(gdb) up
#1  0x00007fffb4e57318 in pmix_iof_read_local_handler (unusedfd=-1, event=1, cbdata=0x10031f89840) at common/pmix_iof.c:1232
1232	        rc = pmix_host_server.push_stdin(&pmix_globals.myid,
(gdb) l
1227	    if (PMIX_PROC_IS_SERVER(&pmix_globals.mypeer->proc_type)) {
1228	        if (NULL == pmix_host_server.push_stdin) {
1229	            /* nothing we can do with this info */
1230	        }
1231	        PMIX_CONSTRUCT_LOCK(&lock);
1232	        rc = pmix_host_server.push_stdin(&pmix_globals.myid,
1233	                                         rev->targets, rev->ntargets,
1234	                                         rev->directives, rev->ndirs,
1235	                                         &bo, opcbfn, &lock);
1236	        PMIX_WAIT_THREAD(&lock);
```

I add a `goto reactivate;` on line 1229 and that seemed to fix the issue.

File: src/common/pmix_iof.c [hunk 1]
```diff
@@ -1215,6 +1223,21 @@ void pmix_iof_read_local_handler(int unusedfd, short event, void *cbdata)
         goto reactivate;
     }
 
+    /* if I am a server, then push this up to my host */
+    if (PMIX_PROC_IS_SERVER(&pmix_globals.mypeer->proc_type)) {
+        if (NULL == pmix_host_server.push_stdin) {
+            /* nothing we can do with this info */
```
Comment (jjhursey) [inline src/common/pmix_iof.c:1229]:
```suggestion
            /* nothing we can do with this info */
            goto reactivate;
```
This seems to fix the IOF issue in CI.

Comment (rhc54) [reply src/common/pmix_iof.c:1229]: Ah...I had forgotten that our simple test servers don't support that function - never occurred to me to look at that path. Thanks!

**Comment by rhc54** [comment]
Rats - still failing the output of 50 instances of `hostname`. I can't reproduce this problem. For some reason, it looks like we are losing output. Only thing I can think of is that somehow the closing of stdin is causing us to somehow think that the application has ended early, thus causing `prterun` to terminate without outputting some of the strings. Is there some way to tell which hosts are missing? In particular, is the missing output from procs local to `prterun` by any chance?

**Comment by rhc54** [comment]
bot:ibm:prrte:retest

**Comment by rhc54** [comment]
bot:ibm:prrte:retest

**Comment by jjhursey** [comment]
CI is a bit backed up still. I tried this manually and it seems to work. But let's wait for CI to run again to make sure
bot:ibm:prrte:retest

**Comment by jjhursey** [comment]
Ah I was testing the wrong thing. Testing the `prun-wrapper` I was able to get a core file:
```
[jjhursey@525277e08e42 (docker) prun-wrapper] ./run.sh < /dev/null 
+ FINAL_RTN=0
+ NUM_NODES=10
+ _HOSTFILE_ARG=
+ '[' x '!=' x/workspace/hostfile.txt ']'
+ _HOSTFILE_ARG='--hostfile /workspace/hostfile.txt'
+ _DASH_HOST_ARG=
+ _DASH_HOST_NUM_NODES=1
+ '[' x '!=' x/workspace/hostfile.txt ']'
++ head -n 1 /workspace/hostfile.txt
+ _HOST_A=525277e08e42
++ tail -n 1 /workspace/hostfile.txt
+ _HOST_B=5f8854a387df
+ _DASH_HOST_ARG='--host 525277e08e42:5,5f8854a387df:5'
+ _DASH_HOST_NUM_NODES=2
+ prterun --map-by ppr:5:node --hostfile /workspace/hostfile.txt hostname
+ tee output-hn.txt
525277e08e42
525277e08e42
525277e08e42
525277e08e42
525277e08e42
c3df0cf86c11
d9a6d1bbbddb
c3df0cf86c11
27077a82eb33
d9a6d1bbbddb
655793c1b79a
c3df0cf86c11
27077a82eb33
c3df0cf86c11
655793c1b79a
c3df0cf86c11
655793c1b79a
655793c1b79a
655793c1b79a
d9a6d1bbbddb
27077a82eb33
d9a6d1bbbddb
d9a6d1bbbddb
27077a82eb33
27077a82eb33
3df4b3c84497
3df4b3c84497
3df4b3c84497
3df4b3c84497
3df4b3c84497
2be33fe0e32c
800f91e8adee
800f91e8adee
2be33fe0e32c
800f91e8adee
2be33fe0e32c
2be33fe0e32c
800f91e8adee
800f91e8adee
2be33fe0e32c
6e3c6494c089
6e3c6494c089
6e3c6494c089
6e3c6494c089
6e3c6494c089
5f8854a387df
5f8854a387df
5f8854a387df
5f8854a387df
++ grep ERROR output-hn.txt
++ wc -l
+ ERRORS=0
+ [[ 0 -ne 0 ]]
++ wc -l output-hn.txt
++ awk '{print $1}'
+ LINES=49
+ [[ 49 -ne 50 ]]
+ echo 'ERROR: Incorrect number of lines of output'
ERROR: Incorrect number of lines of output
+ exit 2
```

```
[jjhursey@525277e08e42 (docker) prun-wrapper] gdb prterun --core core.92226 
(gdb) bt
#0  0x00007fffa0c24614 in push_stdin (dst_name=0x10018137050, data=0x7fffa2ded430 "", sz=0) at iof_hnp.c:242
#1  0x00007fffa3eea0b8 in pmix_server_stdin_push (sd=-1, args=4, cbdata=0x10018127410) at prted/pmix/pmix_server_gen.c:1292
#2  0x00007fffa3ae9e24 in event_process_active_single_queue (base=base@entry=0x10017ec2c90, activeq=0x10017eb8900, 
    max_to_process=max_to_process@entry=2147483647, endtime=endtime@entry=0x0) at event.c:1691
#3  0x00007fffa3aea9b0 in event_process_active (base=0x10017ec2c90) at event.c:1783
#4  event_base_loop (base=0x10017ec2c90, flags=<optimized out>) at event.c:2006
#5  0x000000001000aa50 in prte (argc=6, argv=0x7ffff6cf8df8) at prte.c:1122
#6  0x00000000100039f4 in main (argc=6, argv=0x7ffff6cf8df8) at main.c:13
(gdb) l
237	    }
238	
239	    /* pass the data to the sink */
240	
241	    /* if the daemon is me, then this is a local sink */
242	    if (PMIX_CHECK_PROCID(PRTE_PROC_MY_NAME, &proct->stdinev->daemon)) {
243	        PRTE_OUTPUT_VERBOSE((1, prte_iof_base_framework.framework_output,
244	                             "%s read %d bytes from stdin - writing to %s",
245	                             PRTE_NAME_PRINT(PRTE_PROC_MY_NAME), (int)sz,
246	                             PRTE_NAME_PRINT(&proct->name)));
(gdb) p proct
$1 = (prte_iof_proc_t *) 0x1001812a540
(gdb) p proct->stdinev
$2 = (prte_iof_sink_t *) 0x0
(gdb) 
```

**Comment by rhc54** [comment]
This should pass once https://github.com/openpmix/prrte/pull/845 is committed.

**Comment by rhc54** [comment]
bot:ibm:prrte:retest

**Pull request merged**